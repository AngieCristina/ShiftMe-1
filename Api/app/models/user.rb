class User < ApplicationRecord
  has_many :requester_request, class_name: 'Request', foreign_key: 'requester_id'
  has_many :requested_request, class_name: 'Request', foreign_key: 'requested_id'
end
