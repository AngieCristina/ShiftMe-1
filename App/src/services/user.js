const API_LOGIN_URL = "http://localhost:4000/api/login";
const API_LOGOUT_URL = "http://localhost:4000/api/logout";
const API_USERS_URL = "http://localhost:4000/api/users";

async function createError(response) {
  const { errors } = await response.json();
  const error = new Error(errors.message);
  error.status = response.status;
  error.statusText = response.statusText;
  return error;
}

async function login(user) {
  const response = await fetch(API_LOGIN_URL, {
    method: "POST",
    credentials: "include",
    body: user ? JSON.stringify(user) : "{}",
    headers: {
      "Content-Type": "application/json"
    }
  });
  if (!response.ok) throw createError(response);
  const { name, id, email, rol } = await response.json();
  return { name, email, id, rol};
}

async function logout() {
  const response = await fetch(API_LOGOUT_URL, {
    method: "DELETE",
    credentials: "include"
  });
  if (!response.ok) throw createError(response);
}

async function users() {
  const response = await fetch(API_USERS_URL, {
    credentials: "include",
    headers: {
      "Content-Type": "application/json"
    }
  });
  if (!response.ok) throw createError(response);
  const result = await response.json();

  return result;
}

export { login, logout, users };
