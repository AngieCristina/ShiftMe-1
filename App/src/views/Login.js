/** @jsx jsx */
import { jsx } from "@emotion/core";
import React from "react";
import { Redirect } from "@reach/router";
import { useUser, useUserUpdater } from "../contexts/user";
import { login } from "../services/user";
import { Input, Card, Button } from "../components/Ui";

function Login() {
  const user = useUser();
  const userUpdater = useUserUpdater();
  const [email, setEmail] = React.useState("diego@gmail.com");
  const [password, setPassword] = React.useState("123456");
  const [error, setError] = React.useState(null);

  if (user) return <Redirect to="/" noThrow />;

  function handleEmailChange(event) {
    setEmail(event.target.value);
  }

  function handlePasswordChange(event) {
    setPassword(event.target.value);
  }

  async function handleSubmit(event) {
    event.preventDefault();
    try {
      const { name, id, rol } = await login({ email, password });
      userUpdater({ type: "LOGIN", payload: { name, email, id, rol } });
    } catch (error) {
      setError(error.message);
    }
  }

  return (
    <section css={{ display: "flex", alignItems: "center", height: "100vh" }}>
      <Card css={{ maxWidth: "500px", margin: "1em auto" }}>
        <h1 css={{ margin: "3rem 0", fontSize: "2rem", textAlign: "center" }}>
          ShiftMe
        </h1>
        <div styles={{ marginTop: "3rem", maxWidth: "500px", margin: "auto" }}>
          <form onSubmit={handleSubmit}>
            <fieldset css={{ border: "none", margin: "2em 0", padding: "0" }}>
              <label
                css={{ display: "block", margin: ".5rem 0" }}
                htmlFor="email"
              >
                Email
              </label>
              <Input
                id="email"
                type="email"
                name="email"
                value={email}
                onChange={handleEmailChange}
              />
            </fieldset>
            <fieldset css={{ border: "none", margin: "1em 0", padding: "0" }}>
              <label
                css={{ display: "block", margin: ".5rem 0" }}
                htmlFor="password"
              >
                Password
              </label>
              <Input
                id="password"
                type="password"
                name="password"
                value={password}
                onChange={handlePasswordChange}
              />
              <br />
            </fieldset>
            <Button>Log In</Button>
            {error && (
              <div css={{ color: "red", marginTop: "1rem" }}>
                Error: {error}
              </div>
            )}
          </form>
        </div>
      </Card>
    </section>
  );
}

export default Login;
