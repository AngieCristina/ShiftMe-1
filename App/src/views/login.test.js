import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Login  from "./Login";
import { UserProvider } from "../contexts/user";

test("Login view", () => {
  const user = { name: "", email: "" };
  const setUser = jest.fn();

  const { asFragment } = render(
    <UserProvider user={user} setUser={setUser}>
      <Login />
    </UserProvider>
  );
  expect(asFragment()).toMatchSnapshot();


});
