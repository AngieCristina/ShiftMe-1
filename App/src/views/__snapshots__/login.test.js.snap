// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login view 1`] = `
<DocumentFragment>
  <section
    class="css-13p8oyr-Login"
  >
    <h1
      class="css-1ja91ja-Login"
    >
      Login to the Application
    </h1>
    <div
      styles="[object Object]"
    >
      <form>
        <label
          class="css-1raa8wv-Login"
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          type="email"
          value="diego@gmail.com"
        />
        <label
          class="css-1raa8wv-Login"
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          name="password"
          type="password"
          value="123456"
        />
        <br />
        <button>
          Log In
        </button>
      </form>
    </div>
  </section>
</DocumentFragment>
`;
